JAVA-2288: Improve error reporting when DAO is missing keyspace

tolbertam opened the pull request:
For [JAVA-2288](https://datastax-oss.atlassian.net/browse/JAVA-2288)

---

Motivation:

If no keyspace can be discerned from the session's keyspace, the
entity's default keyspace, or the dao's parameters, the mapper should
preemptively throw a RuntimeException with a helpful message to guide
the user to provide a keyspace.

Modifications:

Added throwIfKeyspaceMissing call in Dao initAsync method to check if
keyspace is set.

Result:

User is now presented with a helpful exception message when keyspace is
not configured.

tolbertam committed:
JAVA-2288: Improve error reporting when DAO is missing keyspace

Motivation:

If no keyspace can be discerned from the session's keyspace, the
entity's default keyspace, or the dao's parameters, the mapper should
preemptively throw a RuntimeException with a helpful message to guide
the user to provide a keyspace.

Modifications:

Added throwIfKeyspaceMissing call in Dao initAsync method to check if
keyspace is set.

Result:

User is now presented with a helpful exception message when keyspace is
not configured.

tolbertam commented on mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java:
```diff
@@ -270,4 +274,19 @@ protected static void throwIfProtocolVersionV3(MapperContext context) {
     }
   }
+
+  protected static <EntityT> void throwIfKeyspaceMissing(
```
The jira ticket called for an error message like:

```
Missing keyspace for userDao(). Suggestions: use SessionBuilder.withKeyspace() when creating your session, specify a default keyspace on User with @Entity(defaultKeyspace), or use a DAO factory method with an @DaoKeyspace parameter
```

I was almost able to match this 1:1 except for providing the dao method name.  Fortunately that shows up in the stack trace.  I think we could probably add some context to resolve it, but wasn't sure how important that was.

olim7t replied:
👍 if it appears in the stack trace that should be clear enough.

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:
```diff
@@ -417,6 +417,7 @@ protected ClassName getPrincipalTypeName() {
       String fieldName = entry.getValue();
       // - create an instance
       initAsyncBuilder.addStatement("$1T $2L = new $1T(context)", fieldTypeName, fieldName);
+      initAsyncBuilder.addStatement("throwIfKeyspaceMissing($L, context.getSession())", fieldName);
```
One thought is that in theory, a user could only specify Dao methods that don't require a keyspace, like `@GetEntity` and `@SetEntity`, we could choose not to fail in this case, but not sure if it's worth worrying about?

olim7t replied:
We could move the check to each entity helper method, we have the context there so I think we have everything we need. Another advantage of that is that the check would also kick in when you call a helper method from a query provider.

We can introduce an `EntityHelperBase` parent class to factor the method.

tolbertam replied:
yeah this seems easy enough to do.  I had initially went about it this way, but seemed like validating in each of these methods was less simple than validating it on entity helper creation. I'll change it.

tolbertam replied:
Adding `EntityHelperBase` also eliminates some boilerplate in the generated code, so that looks like a good change overall.

olim7t commented on mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java:
```diff
@@ -285,4 +289,19 @@ protected static void throwIfProtocolVersionV3(MapperContext context) {
+              "Missing keyspace. Suggestions: use SessionBuilder.withKeyspace() "
+                  + "when creating your session, specify a default keyspace on %s with @%s"
+                  + "(defaultKeyspace), or use a @%s method with a @%s parameter",
+              helper.getEntityClass(),
```
In similar messages (see `replaceKeyspaceAndTablePlaceholders`), I've used the entity's simple name. My reasoning was that it would be pretty obvious from the context which class it is.
I don't mind either way but we should be consistent, if you think it's worth having the package we'll modify the other message.

tolbertam replied:
Fixed that to use `getSimpleName()` :+1:

olim7t reviewed (changes requested):
LGTM, but I think doing the check at the entity helper level (per my other comment) is not too much extra work and would be more precise.

tolbertam committed:
Move validation in EntityHelper methods instead of DaoBase

Also changes generated EntityHelpers to extend EntityHelperBase.

tolbertam committed:
Remove accidental non-null

olim7t commented on mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/entity/EntityHelperBase.java:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  @Override
+  public CqlIdentifier getTableId() {
```
Missing `@NonNull` annotation.

olim7t commented on mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/entity/EntityHelperBase.java:
```diff
@@ -0,0 +1,75 @@
+    this(context, null, tableName);
+  }
+
+  protected EntityHelperBase(MapperContext context, String keyspaceName, String tableName) {
```
Maybe name those parameters `defaultKeyspaceName` and `defaultTableName`.

olim7t commented on mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/entity/EntityHelperBase.java:
```diff
@@ -0,0 +1,75 @@
+          context.getKeyspaceId() != null
+              ? context.getKeyspaceId()
+              : CqlIdentifier.fromCql(keyspaceName);
+    }
```
I'm generally not a big fan of chained ternary operators, but in this case this might be a bit easier to follow:
```java
    this.keyspaceId =
        context.getKeyspaceId() != null
            ? context.getKeyspaceId()
            : (keyspaceName == null ? null : CqlIdentifier.fromCql(keyspaceName));
```

olim7t reviewed (approved):
👍

tolbertam committed:
Qualify keyspace in StatementAttributesIT

This test uses simulacron which doesn't appear to properly respond to
USE keyspace requests.

tolbertam committed:
Feedback

tolbertam commented:
CI looks happy after recent changes, mergin' away
